src/js/sonifier.js

```diff
@@ -0,0 +1,402 @@
+    // execute a sonification
+
+    floe.chartAuthoring.sonifier.startSonification = function(that) {
+        if(!that.model.isPlaying) {
+            // Fire the start event
+
+            that.events.onSonificationStarted.fire();
+
+            flockingEnvironment.start();
+
+            fluid.defaults("floe.chartAuthoring.dataPianoBand", {
+                gradeNames: ["floe.chartAuthoring.electricPianoBand"],
+
+                components: {
+                    scheduler: {
+                        type: "flock.scheduler.async",
+                        options: {
+                            synthContext: "floe.chartAuthoring.electricPianoBand"
+                        }
+                    }
+                }
+            });
+
+            var dataPianoBand = floe.chartAuthoring.dataPianoBand();
+
+            // Add the synth to the model
+            that.applier.change("synth", dataPianoBand);
+
+            var sonifiedData = that.model.sonifiedData;
+            // Copy the sonification definition into the queue
+            that.applier.change("sonificationQueue",sonifiedData);
+        }
+    };
+
+    // Passed a sonified dataset, this function + playDataAndQueueNext acts recursively
+    // to loop through the dataset, play a voice label + sonified data, then
+    // schedule the next play event based on the timing
+
+    // The "delay" variable manages the time to elapse (in seconds) before
+    // beginning the voice label + sonification of the particular datapoint.
+    // This is required for scheduling with the Flocking-based synth generator,
+    // which plays asynchronously without callbacks, while voice events have
+    // callbacks to indicate their speech has completed
+    //
+    // The basic challenge of scheduling that this approach handles is:
+    // - we can calculate in advance how long the sonification of a data point
+    // takes by summing note duration, but we can't receive event notification
+    // when a sonification completes
+    // - we can fire an event when a voice label read completes, but can't know
+    // in advance how long it will take to read the label
+    floe.chartAuthoring.sonifier.processSonificationQueue = function(delay, noGap, that) {
+        that.applier.change("isPlaying", true);
+        var gapDuration = that.options.playbackOptions.gapDuration;
+
+        var sonificationQueue = that.model.sonificationQueue;
+
+        var synth = that.model.synth;
+
+        var currentData = sonificationQueue[0];
+
+        var textToSpeech = that.textToSpeech;
+
+        // Schedule the next voice label, accounting for both the variable-length
+        // delay (the time to play the preceding sonification) and the fixed-length
+        // gap
+
+        // We shouldn't use the gap if this is the first call for a dataset
+        var pause = noGap? delay : delay+gapDuration;
+        if(fluid.textToSpeech.isSupported()) {
```

Comment:
Factor this out into a listener in a "speech-aware" grade which listens to a suitable event rather than embedding it in the core implementation in an "if" block